### PR TITLE
Add padding as object option to fitBounds

### PR DESCRIPTION
**Pull request by GunnarSturla**
This adds an option to set the padding as an object with `top`, `bottom`, `right`, and `left` properties to set different padding for each side when calling the `fitBounds` method. This is useful if you have UI overlaying on one side of the map and you want to have the bounds fill up the rest of the available area.

The api is similar to the way [Mapbox-gl-js's fitBounds](https://www.mapbox.com/mapbox-gl-js/api/#map#fitbounds) method works.

I couldn't find a good example of how this kind of object would be documented, so I ended up describing it inline. I hope that's clear enough, but if there's another way you'd prefer, I'll update accordingly.

Thanks for a great set of tools around mapping and let me know if there's anything I can improve!

### CODE REVIEW

File: src/fit-bounds.js [hunk 1]
```diff
@@ -22,6 +36,27 @@ export default function fitBounds({
...
+    const p = padding;
+    padding = {
+      top: p,
+      bottom: p,
+      left: p,
+      right: p
+    };
+  } else {
+    // Make sure all the required properties are set
+    assert(typeof padding.top === 'number' &&
```
Comment (ibgreen) [inline src/fit-bounds.js]: Nit: We typically use `Number.isFinite` so maybe follow that for consistency.

**Comment by luskin** [comment]
This would be a great improvement, I will be including these changes in my project until they (hopefully) merge, thanks @GunnarSturla

**Comment by luskin** [comment]
Nevermind, this doesn't seem to work as advertised. Adding 100 does not actually pad it 100px on screen on that side it actually moves the map all over the world.

**Comment by GunnarSturla** [comment]
@ibgreen Thanks for your comments, I've fixed the issues you brought up

On the other hand, whoa! @luskin is right, I'm shifting the center of viewport by 100 "pixels" on the Web Mercator plane, which is not what I had intended. I will have to convert the "screen pixels" to "plane pixels" before shifting the center.

I'll have to look at it tomorrow, but thanks @luskin for trying it out and letting me know of the problem!

**Comment by luskin** [comment]
@GunnarSturla looking forward to a patch, with more and more people using floating side bars/cards over maps it seems like a necessary feature going forward.

File: src/fit-bounds.js [hunk 2]
```diff
@@ -37,12 +69,12 @@ export default function fitBounds({
     Math.abs(se[1] - nw[1])
   ];
   const center = [
-    (se[0] + nw[0]) / 2,
-    (se[1] + nw[1]) / 2
+    (se[0] + nw[0]) / 2 + lateralOffset,
+    (se[1] + nw[1]) / 2 + verticalOffset
```
Comment (Pessimistress) [inline src/fit-bounds.js]: `lateralOffset` and `verticalOffset` are in pixels at the new zoom.
`se` and `nw` are in pixels at the old zoom.

**Comment by GunnarSturla** [comment]
Thanks for your help @Pessimistress, I've been trying to figure this thing out, but I haven't had too much time to wrap my head around it all, so it's taken its time.

`lateralOffset` and `verticalOffset`* are screen pixels (the difference between opposite paddings), and if I understand this correctly, `nw` and `se` are the locations of the corners of the bounding box on the Web Mercator plane (in pixels or units). It took me a while to figure out what `scaleX` and `scaleY` were, but it looks like it's screen pixels per unit on the Web Mercator plane. I've used that scale to convert the 'screen pixels' to 'Web Mercator pixels', and offset the center correctly.

All the calculations are done using the `viewport` set on line 54, at zoom level 0. Even though the pixel/unit values might be different at different zoom level, it's just used to figure out where the center of the map should be, and since it's then unprojected to LngLat, it should all cancel out.

* I've changed the variable names for `verticalOffset` and `lateralOffset` to `offsetX` and `offsetY` in order to match the `scaleX`/`Y` names.

**Comment by Pessimistress** [comment]
@GunnarSturla No problem, I appreciate the contribution.

A few other things I realized during review, not necessarily need to be addressed in this PR:
- We haven't been checking if padding is larger than the viewport size. It will be nice if an error is thrown here rather than failing to render later.
- There needs to be some sanity check for the test cases (either checking projected corner positions or compare with Mapbox)

**Comment by Pessimistress** [comment]
Published v5.0.2

**Comment by GunnarSturla** [comment]
Awesome! Thanks @Pessimistress!

**Comment by WillEngler** [comment]
Heads up: it looks like the documentation that this PR added to fitBounds is not yet showing up on the GH pages [documentation site](http://uber-common.github.io/viewport-mercator-project/#/documentation/api-reference/web-mercator-utils). I was thinking of making an issue to ask if I could add the feature before I checked the source. Might be time to re-run the site generator? 

(In any event, I'm a big fan of these libraries - thank you maintainers!)
